c7n/resources/s3.py

```diff
@@ -2494,3 +2500,119 @@ def process_bucket(self, bucket):
+                  # enabled: true <------ optional (true by default)
+                    crypto: aws:kms
+                    key: alias/some/alias/key
+
+              - name: s3-enable-default-encryption-aes256
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                  # crypto: AES256 <----- optional (AES256 by default)
+                  # enabled: true <------ optional (true by default)
+
+              - name: s3-disable-default-encryption
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                    enabled: false
+    """
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'type': {'enum': ['set-bucket-encryption']},
+            'enabled': {'type': 'boolean'},
+            'crypto': {'enum': ['aws:kms', 'AES256']},
+            'key': {'type': 'string'}
+        },
+        'dependencies': {
+            'key': {
+                'properties': {
+                    'crypto': {'pattern': 'aws:kms'}
+                },
+                'required': ['crypto']
+            }
+        }
+    }
+
+    permissions = ('s3:PutEncryptionConfiguration',
+                's3:GetEncryptionConfiguration',
+                'kms:ListAliases')
+
+    def process(self, buckets):
+        with self.executor_factory(max_workers=3) as w:
+            results = []
+            enabled = self.data.get('enabled', True)
+            if enabled:
+                SSEAlgorithm = self.data.get('crypto', 'AES256')
+                config = {'Rules': [
+                    {
+                        'ApplyServerSideEncryptionByDefault': {
+                            'SSEAlgorithm': SSEAlgorithm
+                        }
+                    }
+                ]}
+                key = self.data.get('key')
+                if key:
+                    config['Rules'][0]['ApplyServerSideEncryptionByDefault']['KMSMasterKeyID'] = key
+            else:
+                config = None
+            futures = {w.submit(self.process_bucket, b, config) : b for b in buckets}
+            for future in as_completed(futures):
+                if future.exception():
+                    bucket = futures[future]
+                    self.log.error('error enabling bucket encryption: %s\n%s',
+                                   bucket['Name'], future.exception())
+                results += filter(None, [future.result()])
+            return results
+
+    def process_bucket(self, bucket, config):
+        s3 = bucket_client(local_session(self.manager.session_factory), bucket)
+        if config:
+            key = config['Rules'][0]['ApplyServerSideEncryptionByDefault'].get('KMSMasterKeyID')
+            if key and key.startswith('alias'):
+                # Unfortunately need to pull aliases per bucket as buckets can be in any region.
```

Comment:
this can go into a helper, also instead of fetching the kms aliases yourself reusing the resource manager for aliases will help utilize the cache, as is your effectively fetching the same list for every single bucket.